## 1. 트리 앙상블 (Tree Ensemble)

> **정의:** 트리 앙상블(Tree Ensemble)은 여러 개의 결정 트리(Decision Tree)를 결합하여 예측 성능을 향상시키는 머신러닝 기법입니다.

### 단일 결정 트리의 특징

- 구조가 직관적이어서 **해석이 쉽습니다.**
- 하지만 **훈련 데이터에 과하게 맞춰지는 과적합(Overfitting)** 이 발생하기 쉽습니다.
- 데이터가 조금만 바뀌어도 모델 구조가 크게 달라지는 **불안정성(High Variance)** 이 있습니다.

### 앙상블의 해결 방법

- 여러 개의 결정 트리를 학습시켜 **각 모델의 예측을 결합**합니다.
- 각 트리는 조금씩 다른 데이터나 조건으로 학습됩니다.
- 최종 결과는 다음 방식으로 결합합니다.
    - **분류(Classificaiton):** 투표(Voting)
    - **회귀(Regression):** 평균(Averaging)

### 핵심 원리

- 여러 모델의 예측을 결합하면 **개별 모델의 오류가 서로 보완됩니다.**
- 그 결과 **예측이 더 안정적이고 일반화 성능이 좋아집니다.**

 ---
이러한 **트리 기반 앙상블 모델**은 특히 **표 형태의 데이터를 다루는 문제에서 매우 강력한 성능**을 보입니다.  
머신러닝에서는 데이터의 형태에 따라 **정형 데이터**와 **비정형 데이터**로 구분하며,  
각 데이터 유형에 따라 잘 맞는 알고리즘이 다릅니다.

## 2. 정형 데이터와 비정형 데이터

| **구분** | **정형 데이터 (Structured)** | **비정형 데이터 (Unstructured)** |
| :--- | :--- | :--- |
| **형태** | 표 형태 (열/행 구조 명확) | 텍스트, 이미지, 오디오 등 |
| **저장** | CSV, DB, Excel | 파일 (문서, 이미지, 음원) |
| **대표 알고리즘** | **트리 기반 앙상블** | **신경망 (딥러닝)** |
| **특징** | 데이터 타입/구조가 미리 정의됨 | 구조가 일정하지 않음 |

- **정형 데이터**

    - 표 형태로 정리된 데이터입니다.
    - 예: 고객 정보, 금융 데이터, 판매 기록 등
    - 이런 데이터에서는 **트리 기반 앙상블(Random Forest, Gradient Boosting 등)** 이 매우 강력한 성능을 보입니다.

- **비정형 데이터**

    - 구조가 일정하지 않은 데이터입니다.
    - 예: 이미지, 텍스트, 음성, 영상 등
    - 이러한 데이터는 보통 **신경망(딥러닝)** 이 더 효과적으로 처리합니다.

 ---

## 3. 랜덤포레스트 (Random Forest)

랜덤 포레스트는 **트리 앙상블의 대표적인 알고리즘**입니다.  
여러 개의 결정 트리를 독립적으로 학습시킨 후 결과를 결합하여 예측합니다.

### 3-1. 개념

랜덤 포레스트는 여러 개의 결정 트리를 만들어 각 트리의 예측을 합쳐 **더 안정적인 결과를 만드는 앙상블 모델**입니다.

### 3-2. 동작 단계

1. 부트스트랩 샘플 생성 (Bootstrap Sampling)

    - 원본 데이터에서 **중복을 허용**하여 랜덤하게 샘플을 추출합니다.
    - 이 때 선택되지 않은 데이터는 **OOB (Out-of-Bag)** 라고 합니다.

2. 특성 랜덤 선택

    - 노드를 분할할 때 **모든 특성을 사용하지 않고 일부 특성만 랜덤하게 선택**합니다.  
    예) 전체 특성 : 길이, 무게, 높이, 너비  
    랜덤 선택 : 길이, 너비만 사용

3. 예측

    - 각 트리가 예측한 결과를 결합하니다.
    - **분류 문제:** 다수결 투표 (Voting), **회귀 문제:** 평균 (Averaging)

4. OOB(Out-of-Bag) 평가

    - 부트스트랩 샘플링에서 선택되지 않은 **OOB 데이터**를 사용해 별도의 검증 세트 없이 모델 성능을 평가할 수 있습니다.

---

## 정리

트리 앙상블은 **여러 개의 결정 트리를 결합하여 성능을 높이는 머신러닝 기법**이며, 특히 **정형 데이터에서 매우 강력한 성능**을 보입니다.

대표적인 알고리즘이 **랜덤 포레스트(Random Forest)** 이며 부트스트랩 샘플링과 특성 랜덤 선택을 통해 **서로 다른 트리들을 만들고 그 결과를 결합하여 예측합니다.**

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

wine = pd.read_csv('data/wine.csv')

data = wine[['alcohol', 'sugar', 'pH']].to_numpy()
target = wine['class'].to_numpy()

train_input, test_input, train_target, test_target = train_test_split(data, target, test_size=0.2, random_state=42)

In [2]:
from sklearn.model_selection import cross_validate
from sklearn.ensemble import RandomForestClassifier
# import os
# os.makedirs(r"C:\temp\joblib", exist_ok=True)
# os.environ["JOBLIB_TEMP_FOLDER"] = r"C:\temp\joblib"  # 영문 경로

# RandomForestClassifier
# 여러 개의 결정 트리를 학습한 뒤, 각 트리의 예측 결과를 결합하여 최종 예측을 수행하는 트리 기반 앙상블 모델
# 분류 무제에서는 다수결(Voting)을 사용합니다.

# n_estimators=100 : 생성할 결정 트리의 수
rf = RandomForestClassifier(n_jobs=-1, random_state=42)

# xxxx_score : 각 폴드에서 xxx데이터로 계산한 점수
# fit_time: 학습 시간
# score_time : 점수 계산 시간

scores = cross_validate(rf,
                        train_input,
                        train_target,
                        return_train_score=True, # train 점수도 함께 계산
                        n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9973541965122431 0.8905151032797809


훈련 점수(train_score) > 검증 점수 (test_score) 차이가 0.10으로 약 10% 차이가 납니다. 과대 적합 패턴입니다

In [3]:
rf.fit(train_input, train_target)
print(rf.feature_importances_)
# 이전의 결정 트리에서 결정 중요도가 당도에 몰려있었습니다.
# 이번에는 그보다 좀 더 고르게 변했습니다.
# 특성 랜덤 선택 때문입니다.
# 많은 특성이 훈련에 기여할 기회를 얻게 되는 겁니다.

[0.23167441 0.50039841 0.26792718]


 ---
과대 적합 해결 방법
- max_depth=10, min_samples_leaf=5 적용하기

In [4]:
rf = RandomForestClassifier(n_jobs=-1, random_state=42, max_depth=10, min_samples_leaf=5)

scores = cross_validate(rf,
                        train_input,
                        train_target,
                        return_train_score=True,
                        n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9115357268111659 0.8783941659880062


In [5]:
rf.fit(train_input, train_target)
print(rf.feature_importances_)

[0.17494464 0.60018378 0.22487159]


 ---
oob_score=True 옵션 주기

In [6]:
# 각 트리는 자신이 학습에 사용하지 않은 OOB 데이터를 이용해 평가됩니다.
# 이렇게 계산된 점수를 평균하여 OOB 점수를 구합니다.
# OOB 데이터는 별도의 검증 세트처럼 모델 성능을 추정하는 역할을 합니다.
rf = RandomForestClassifier(oob_score=True, n_jobs=-1, random_state=42)

rf.fit(train_input, train_target)

print(rf.oob_score_)

0.8934000384837406


 ---
## 4. 엑스트라 트리 (Extra Trees)

- **개념:**  
랜덤 포레스트와 매우 유사한 트리 기반 앙상블 모델로, 여러 개의 결정 트리를 만들어 그 결과를 결합하여 예측합니다.

- **특징**

    - **부트스트랩 샘플을 사용하지 않고 전체 데이터로 트리를 학습합니다.**
    - 노드를 분할할 때 **최적의 분할 기준을 탐색하지 않고 무작위 분할 지점(random split)** 을 선택합니다.

- **장점**

    - 분할 기준을 찾는 계산 과정이 줄어들어 **학습 속도가 매우 빠릅니다.**
    - 분할이 무작위로 이루어지기 때문에 **트리 간 다양성이 커지고 과적합을 강하게 억제할 수 있습니다.**

- **단점**

    - 분할 기준을 최적으로 찾지 않기 때문에 **데이터에 따라 랜덤 포레스트보다 성능이 낮을 수 있습니다.**

In [7]:
from sklearn.ensemble import ExtraTreesClassifier

et = ExtraTreesClassifier(n_jobs=-1, random_state=42)
scores = cross_validate(et, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9974503966084433 0.8887848893166506


In [8]:
et.fit(train_input, train_target)
print(et.feature_importances_)

[0.20183568 0.52242907 0.27573525]


 ---
## 부스팅 (Boosting)

부스팅은 여러 개의 **단순한 모델(Simple Model)** 을 **순차적으로 학습**시키면서 이전 모델이 만든 **오류를 다음 모델이 보완하도록 하는 앙상블 방법**입니다.

- **작동 원리**

    - 첫 번째 모델이 예측을 수행합니다.
    - 이때 틀린 데이터(오류)가 발생합니다.
    - 다음 모델은 **이 오류를 더 잘 맞추도록 학습**합니다.
    - 이러한 과정을 반복하면서 모델의 성능이 점점 개선됩니다.

- **특징**

    - 모델들이 **순차적으로 학습됩니다.**
    - 이전 모델의 **오류를 보완하면서 점점 더 성능이 향상됩니다.**
    - 여러 개의 **단순한 모델을 결합하여 강렬한 모델을 만듭니다.**

- **대표적인 Boosting 알고리즘**

    - Gradient Boosting
    - XGBoost
    - LightGBM

---

## 5. 그레이디언트 부스팅 (Gradient Boosting)

이전 모델이 마든 오차(잔차)를 다음 모델이 보완하는 방식으로  
단순한 구조의 결정 트리를 **순차적으로 추가하며 모델을 개선하는 앙상블 기법**입니다.

- **작동 원리**

    - 첫 번째 트리가 예측을 수행합니다.
    - 이때 발생한 오차(잔차)를 계산합니다.
    - 다음 트리는 이 **오차를 줄이도록 학습**합니다.
    - 각 트리는 결정 트리 방식으로 학습되며 **노드 분할 시 불순도를 가장 많이 줄이는 최적 분할점(best split)을 찾습니다.**
    - 이러한 과정을 반복하면서 모델을 점점 개선합니다.

- **기울기(Gradient)의 의미**

    - 손실 함수(loss)를 줄이기 위해 손실이 가장 빠르게 감소하는 방향(기울기)으로 모델을 업데이트합니다.

- **성능**

    - 정형 데이터 문제에서 **매우 강력한 성능**을 보입니다.
    - 하지만 모델이 **순차적으로 학습**되기 때문에 학습 속도가 비교적 느릴 수 있습니다.

- **대표적인 구현 라이브러리**

    - XGBoost : 속도와 성능을 대폭 개선한 라이브러리
    - LightGBM : 대용량 데이터에 최적화된 빠른 부스팅
    - CatBoost : 범주형 데이터를 처리하는 데 탁월한 모델

 ---
GradientBoostingClassifier

- 이전 모델의 오차(잔차)를 줄이도록 작은 결정 트리를 순차적으로 추가하는 트리 기반 앙상블 모델입니다.

- 기본적으로 깊이 3인 결정 트리(max_depth=3)를 100개 사용합니다.

- 손실 함수를 줄이기 위해 손실 함수의 기울기(gradient)를 이용하여 새로운 트리를 순차적으로 추가하면서 모델을 개선합니다.

- 분류 문제에서는 기본적으로 로지스틱 손실 함수(log_loss)를 사용합니다.

- 즉, 이전 모델이 틀린 부분을 다음 모델이 계속 보완하면서 전체 모델의 성능을 점점 향상시키는 방식입니다.

In [9]:
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(random_state=42)

scores = cross_validate(gb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.8881086892152563 0.8720430147331015


거의 과대 적합이 되지 않았습니다.

 ---
학습률 (learning_rate=0.05)를 감소시키고 트리의 갯수(n_estimators=300)를 늘려보도록 하겠습니다.

In [10]:
gb = GradientBoostingClassifier(n_estimators=300,   # 기본값 100
                                learning_rate=0.05, # 기본값 0.1
                                random_state=42)
scores = cross_validate(gb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.8920531592891467 0.8733902791145333


In [11]:
gb.fit(train_input, train_target)
print(gb.feature_importances_)

[0.12676547 0.73346135 0.13977318]


이전의 결정 트리에서 특성 중요도는 [0.12345626 0.86862934 0.0079144]

이전 결정 트리(Decision Tree) 에선느 세 번째 특성의 중요도가 0.0079로 거의 사용되지 않았습니다.

반면, Gradient Boosting 에서는 같은 특성의 중요도가 0.1398로 크게 증가했습니다.

결정 트리는 하나의 강력한 특성에 의존하는 경향이 강하고,  
Gradient Boosting은 주요 특성으로 전체 패턴을 먼저 학습한 뒤,  
이전 모델이 놓친 오차를 보정하기 위해 나머지 특성들을 적극적으로 활용하며 학습합니다.

---

## 6. 히스토그램 기반 그레이디언트 부스팅

### (1) 그레이디언트 부스팅 속도 문제

일반적인 그레이디언트 부스팅은 트리를 분할할 때 **가능한 모든 분할 지점을 검사하여 최적의 분할을 찾습니다.**

즉, 각 특서으이 값들을 하나씩 비교하면서 어디에서 노드를 분할할지 계산해야 합니다.

이 과정이 **모든 노드와 모든 트리에서 반복되기 때문에** 데이터가 많아질수록 **학습 시간이 크게 증가**합니다.

### (2) 해결 방법 : 히스토그램 기반 방법

이러한 **속도 문제를 해결하기 위해 등장한 방법**이 **히스토그램 기반 그레이디언트 부스팅(Histogram-based Gradient Boosting)** 입니다.

이 방법은 입력 데이터를 **구간 (Bin)** 으로 나누어 계산량을 줄입니다.

### (3) 작동 원리

1. 각 특성 값을 일정한 **구간 (Bin)** 으로 나눕니다.

2. 각 구간에 속하는 데이터 개수를 이용해 **히스토그램(Histogram)** 을 만듭니다.

3. 분할 지점을 찾을 때 **모든 개별 값을 비교하는 대신 구간 단위로 계산**합니다.

    보통 특성 값은 약 **256개의 구간(bin)** 으로 나누어 사용합니다.

### (4) 효과

히스토그램 기반 방법을 사용하면

- 개별 데이터를 모두 비교하지 않아도 되고
- 구간 단위로 계산할 수 있기 때문에

다음과 같은 장점이 있습니다.

- **학습 속도가 매우 빨라짐**
- **메모리 사용량 감소**
- **대용량 데이터 처리에 유리**

---

### 모델 선택 가이드

상황에 따라 다음과 같이 모델을 선택할 수 있습니다.

**안정적이고 기본적인 모델이 필요할 때  
→ 랜덤 포레스트 (Random Forest)**

**빠른 학습 속도가 중요할 때  
→ 엑스트라 트리 (Extra Tree)**

**대용량 데이터에서 빠른 부스팅이 필요할 때  
→ 히스토그램 기반 부스팅 (HistGradientBoosting, LightGBM)**

**최상의 예측 성능을 얻고 싶을 때  
→ 그레이디언트 부스팅 계열 (XGBoost, LightGBM, CatBoost 등)**

In [12]:
from sklearn.ensemble import HistGradientBoostingClassifier

hgb = HistGradientBoostingClassifier(random_state=42)

scores = cross_validate(hgb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9321723946453317 0.8801241948619236


히스토그램 기반 그레이디언트 부스팅은 자체적으로 특성 중요도를 제공하지 않습니다.

특성 중요도를 계산하기 위해 permutation_importance() 함수를 사용하겠습니다.

In [ ]:
from sklearn.inspection import permutation_importance

hgb.fit(train_input, train_target)

# permutation_importance는 각 특성 값을 무작위로 섞어(permutation) 모델 성능이 얼마나 감소하는지 측정하여 특성의 중요도를 계산하는 방법입니다.
# 값이 클수록 해당 특성이 모델 예측에 더 중요한 특성임을 의미합니다.
result = permutation_importance(hgb, train_input,
                                train_target,
                                n_repeats=10,
                                random_state=42, n_jobs=-1)
print(result.importances_mean)

# 두 번째 특성을 섞으면 0.2344 정도 모델의 성능이 감소한다.
# 따라서 해당 특성이 가장 중요한 특성인 것이다.

[0.08876275 0.23438522 0.08027708]


In [14]:
# 테스트 세트에 대한 permutation_importance()
result = permutation_importance(hgb, test_input, test_target, n_repeats=10,
                                random_state=42, n_jobs=-1)
print(result.importances_mean)

[0.05969231 0.20238462 0.049     ]


In [15]:
# 테스트 세트에 대한 score()
hgb.score(test_input, test_target)

0.8723076923076923

 ---
### XGBoost

XGBoost(eXtreme Gradient Boosting)는 **Gradient Boosting 알고리즘을 개선하여 만든 고성능 트리 기반 앙상블 라이브러리**입니다.

In [16]:
%pip install xgboost

   ---------------------------------------- 0.0/101.7 MB ? eta -:--:--
   - -------------------------------------- 3.9/101.7 MB 26.0 MB/s eta 0:00:04
   ---- ----------------------------------- 11.0/101.7 MB 29.9 MB/s eta 0:00:04
   ------- -------------------------------- 19.9/101.7 MB 34.0 MB/s eta 0:00:03
   ----------- ---------------------------- 28.3/101.7 MB 35.9 MB/s eta 0:00:03
   -------------- ------------------------- 36.7/101.7 MB 36.4 MB/s eta 0:00:02
   ----------------- ---------------------- 43.8/101.7 MB 36.2 MB/s eta 0:00:02
   ------------------- -------------------- 50.6/101.7 MB 35.8 MB/s eta 0:00:02
   ----------------------- ---------------- 59.8/101.7 MB 36.6 MB/s eta 0:00:02
   --------------------------- ------------ 70.3/101.7 MB 38.3 MB/s eta 0:00:01
   ------------------------------- -------- 80.5/101.7 MB 39.5 MB/s eta 0:00:01
   ----------------------------------- ---- 91.2/101.7 MB 40.4 MB/s eta 0:00:01
   -------------------------------------- - 98.3/1

 ---
### XGBClassifier

- **XGBClassifier**는 XGBoost 라이브러리에서 제공하는 **분류(Classification) 모델 클래스**입니다.

- 즉, **XGBoost 알고리즘을 이용하여 분류 문제를 해결하기 위해 구현된 모델**입니다.

- 이 모델은 **Gradient Boosting 방식**을 사용하여 이전 모델의 오차를 줄이도록 **결정 트리를 순차적으로 추가**하면서 학습합니다.

- 주로 **정형 데이터(tabular data)** 분류 문제에서 높은 성능을 보입니다.

In [17]:
from xgboost import XGBClassifier

xgb = XGBClassifier(tree_method='hist', random_state=42)
scores = cross_validate(xgb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9567059184812372 0.8783915747390243


 ---
#### LightGBM

- **LightGBM(Light Gradient Boosting Machine)** 은  
Gradient Boosting 알고리즘을 기반으로 만든 **고속·고성능 머신러닝 라이브러리**입니다.

- **히스토그램 기반 학습(Histogram-based Learning)** 을 사용하여  
데이터를 여러 구간(bin)으로 나누어 계산하기 때문에 **학습 속도가 매우 빠릅니다.**

- 기존 Gradient Boosting보다 **메모리 사용량이 적고 대규모 데이터 학습에 유리**합니다.

- 주로 **정형 데이터(tabular data)** 문제에서 매우 높은 성능을 보입니다.

In [18]:
%pip install lightgbm

   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   ---------------------------------------- 1.5/1.5 MB 15.2 MB/s  0:00:00
Note: you may need to restart the kernel to use updated packages.


 ---
### LGBMClassifier

- **LGBMClassifier** 는 LightGBM 라이브러리에서 제공하는 **분류(classification) 모델 클래스**입니다.

- LightGBM 알고리즘을 **분류 문제에 사용할 수 있도록 구현한 모델**입니다.

In [19]:
from lightgbm import LGBMClassifier

lgb = LGBMClassifier(random_state=42)
scores = cross_validate(lgb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.935828414851749 0.8801251203079884
